File: quickload-core/src/main/java/org/quickload/spi/ParserPlugin.java
package org.quickload.spi;

import org.quickload.config.ConfigSource;
import org.quickload.config.TaskSource;
import org.quickload.channel.FileBufferInput;
import org.quickload.channel.PageOutput;

public interface ParserPlugin
{
    public TaskSource getParserTask(ExecTask exec, ConfigSource config);

    public void runParser(ExecTask exec,
            TaskSource taskSource, int processorIndex,
            FileBufferInput fileBufferInput, PageOutput pageOutput);
}

File: lib/quickload/bridge/parser_plugin.rb
module QuickLoad
  module Bridge
    require 'quickload/java/imports'
    require 'quickload/bridge/data_source'
    require 'quickload/bridge/schema'
    require 'quickload/file_reader'
    require 'quickload/record_writer'

    class ParserPluginBridge < Java::BasicParserPlugin
      def getBasicParserTask(proc, config)
        config = DataSourceBridge.wrap(config)
        task = TaskConfig.new(proc)
        configure(config, task)
        proc.setSchema(Bridge::Schema.new(task.columns).to_java)
        DataSourceBridge.to_java_task_source(task)
      end

      def runBasicParser(proc, taskSource, processorIndex, fileBufferInput, pageOutput)
        task = TaskConfig.new(proc, DataSourceBridge.wrap(taskSource))
        file_reader = FileReader.new(fileBufferInput)
        record_writer = RecordWriter.new(task, pageOutput)
        process(task, processorIndex, file_reader, record_writer)
        nil
      end
    end

    ## TODO
    #class LineParserPlugin < Java::BasicParserPlugin
    #end

    ## TODO
    #class TextParserPlugin < Java::BasicParserPlugin
    #end

  end
end
